File: async_sinatra/data_access.rb
class DataAccess
  
  def self.begin(&block)
    block.call DataAccess.new
  end
  
  def cards_in_set(name, &block)
    set_name = escape(name)
    redis.smembers("set-cards-#{set_name}") do |keys|
      card_keys = keys.map { |key| "card-#{key}" }
      objects_for_keys(card_keys) { |objs| block.call objs }
    end
  end
  
  def close
    redis.close_connection
  end
  
private  
  def redis
    @redis ||= EM::Hiredis.connect
  end
  
  def escape(string)
    string.gsub(/\W/, "-").downcase.gsub(/-+/, '-')
  end
  
  def objects_for_keys(keys, &block)
    if keys.length > 0
      redis.mget(*keys) do |objs|
        #multiple returned objects need to be added to JSON array
        block.call "[#{objs.join(", ")}]"
      end
    else
      block.call "[]"
    end
  end
    
  
end

File: Node.js/data_access.js
var redisLib = require("redis")

DataAccess = {
  init: function(){
    var self = Object.create(DataAccess);
    var redis = redisLib.createClient();
    var escape = function(string){
      return string.replace(/\W/g, "-").toLowerCase().replace(/-+/g, '-');
    };
    var objectsForKeys = function(keys, block){
      if(keys.length <= 0){
        block("[]");
      }else{
        redis.mget(keys, function(e, objs){
          block("[" + objs.join(",") +  "]");
        });
      }
    };
    
    self.allSets = function(block){
      redis.smembers("sets", function(e, set_names){
        var sets = [];
        set_names.forEach(function(name){
          sets.push({id: escape(name), name: name});
        });
        block(sets);
      });
    };
    
    self.cardsInSet = function(name, block){
      setName = escape(name);
      redis.smembers("set-cards-" + setName, function(e, guids){
        var keys = guids.map(function(guid){ return "card-" + guid; });
        objectsForKeys(keys, block);
      });
    };
    
    self.close = function(){
      redis.quit();
    };
    
    return self;
  },
  
  begin: function(block){
    block(DataAccess.init());
  }
}